### PR TITLE
sys.meta_path import mapping layer

### PR DESCRIPTION
**Pull request by smackesey**
This PR injects a `sys.meta_path` `MetaPathFinder` at the top of `dagster/__init__.py`. This maps imports of the form `dagster.XXX` to `dagster._XXX`. This allows for a submodule renaming to support the typed library standard (see #7017) while allowing imports under old module paths to continue working, thus preventing breakage for users who are directly importing dagster submodules for whatever reason.

## Test Plan

A few unit tests added to ensure dynamic import resolution is working.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #7036** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/7036" target="_blank">Graphite</a> 
    * **PR #7017** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/7017" target="_blank">Graphite</a> 
      * **PR #7040** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/7040" target="_blank">Graphite</a>  (this PR)
        * **PR #7039** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/7039" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

File: python_modules/dagster/dagster/_module_alias_map.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                AliasedModuleLoader(fullname, base_spec),
+                origin=base_spec.origin,
+                is_package=base_spec.submodule_search_locations is not None,
+            )
+        else:
+            return None
+
+
+# Key reference to understand the load process:
+#   https://docs.python.org/3/reference/import.html#loading
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_module_alias_map.py:47]: [1]

**Review by alangenfeld — COMMENTED** [review]
This is a bit terrifying. Was there documentation you referred to that inspired this approach? I don't see MetaPathFinder in [1]

**Comment by smackesey** [comment]
This is the master documentation: https://docs.python.org/3/reference/import.html

And these two articles are a little more digestible:

- https://dev.to/dangerontheranger/dependency-injection-with-import-hooks-in-python-3-5hap
- https://github.com/0cjs/sedoc/blob/master/lang/python/importers.md

LMK if you have other questions around how something works. Took me a while to wrap my head around it, the docs could definitely be clearer. I actually started with this stuff before I started working here, I was experimenting with `sys.meta_path` for direct imports of "single-file ops".

Here's a more thorough explanation in terms of this PR's code:

When you import a module, Python tries to resolve it by trying these steps (in order):

- Normalizes any relative path, and breaks the "fully-qualified" module name (e.g. `dagster.core.definitions`) into a sequence of modules with ancestor modules first (`dagster`, `dagster.core`, `dagster.core.defintions`).
- Tries to resolve each of these modules in order, using the following protocol:

1. Checks `sys.modules` (dict cache holding everything already imported keyed by fully-qualified module name)

2. If not in `sys.modules`, iterates over `sys.meta_path`. All entries of `sys.meta_path` inherit from`importlib.abc.MetaPathFinder`. By default `sys.meta_path` contains 3 `MetaPathFinders`: (1) for finding builtin modules; (2) for finding frozen modules; (3) for finding anything on `sys.path`. In this PR we are injecting our `meta_path_finder` in front of the last (`sys.path`) one. Note that the `sys.path` meta path finder contains a fairly complex subsystem for resolving stuff on `sys.path`, we don't need to go into that here.

`MetaPathFinder` has one required method, `find_spec(fullname, path, target)`. `fullname` is the "full" module name, e.g. `dagster.core.definitions`. `path` is the `__path__` attribute of the parent module or `None` for top-level, `target` is only use during re-loading and is irrelevant for purposes of this explanation. Note we have access to `__path__` attribute because the parent is always resolved immediately before the child. (Note that `find_spec` can  do anything you want-- you can use it to generate modules with totally arbitrary logic.)

During iteration, we call `find_spec` for each `MetaPathFinder`, which returns either a `ModuleSpec` or `None`. If `ModuleSpec` stop iterating and go to next stage. If `None`, keep iterating. If you reach the end of `sys.meta_path` and no spec was found , `ModuleNotFoundError`.

3. The `ModuleSpec` object returned from `sys.meta_path` iteration contains metadata about a module and, crucially, info on how to load it encapsulated in a `importlib.abc.Loader` object under the `loader` attribute. Let's inline segments of the comment-linked snippet from the import docs here `Loader` docs here for clarity.

First, we are going to get an actual module object (type `ModuleType`) from our module spec. We first try invoking `loader.create_module` (this can do anything you want). If this method is not defined or returns `None`, then we create the module with default logic (`module = ModuleType(spec.name)`). Notice that `AliasedModuleLoader` above does not define `create_module`. That's because we don't care about the module created at this stage, because we're going to throw it away and replace it with a module loaded from an actually existing path.

```
module = None
if spec.loader is not None and hasattr(spec.loader, 'create_module'):
    # It is assumed 'exec_module' will also be defined on the loader.
    module = spec.loader.create_module(spec)
if module is None:
    module = ModuleType(spec.name)
```

After we have a `ModuleType`, Python sets metadata on the module pulled off the spec (`__spec__`, `__name__`) etc. There appears to be no way to intervene in this process. For that reason, we can't swap out e.g. `dagster.core` for `dagster._core` at the module creation stage, because e.g. `dagster._core.__name__` would be overwritten from `dagster._core` to `dagster.core`.

```
# The import-related module attributes get set here:
_init_module_attrs(spec, module)
```

Finally we are going to load and cache the module in the final conditional. The first three conditions below are irrelevant for our purposes and concern errors (no loader defined), namespace packages, and an old API (`load_module`):

```
if spec.loader is None:
    # unsupported
    raise ImportError
if spec.origin is None and spec.submodule_search_locations is not None:
    # namespace package
    sys.modules[spec.name] = module
elif not hasattr(spec.loader, 'exec_module'):
    module = spec.loader.load_module(spec.name)
    # Set __loader__ and __package__ if missing.
```

Below is what we care about. Notice that python caches the (currently empty) module in `sys.modules` _before_ execution. This is necessary to handle circular imports. Then we `exec_module` the module. This is arbitrary code that can do anything and is intended to mutate the passed-in module. Finally we return the module from the `sys.modules`.

else:
    sys.modules[spec.name] = module
    try:
        spec.loader.exec_module(module)
    except BaseException:
        try:
            del sys.modules[spec.name]
        except KeyError:
            pass
        raise
return sys.modules[spec.name]
```

Finally, here's the `exec_module` from `AliasedModuleLoader`. If we're importing, say, `dagster.core`, all we are doing is importing the base module (e.g. `dagster._core` for `dagster.core`), and then overwriting the `sys.modules` cache entry with the `base_module`. So now `sys.modules["dagster.core"]` and `sys.modules["dagster._core"]` point to exactly the same object. Note that importing of `dagster._core` will be handled like any module on `sys.path`, i.e. by the `sys.path` meta path finder in `sys.meta_path` (it will be passed to AliasedModuleFinder first but `find_spec` will return `None`).

```
    def exec_module(self, _module: ModuleType) -> None:
        base_module = importlib.import_module(self.base_spec.name)
        sys.modules[self.alias] = base_module
```

**Review by alangenfeld — CHANGES_REQUESTED** [review]
Thanks for the very thorough response. Good to have all that information recorded here. 

I am going to put this back in your queue for now, until we are ready to land these changes in accordance to a sequencing and end state that more of the team has signed off on.

**Review by alangenfeld — APPROVED** [review]
🙏🏼

**Pull request merged**